File: worker/src/segments.ts
/**
 * Route segment utilities
 * (Standalone copy of lib/analytics/segments.ts for the worker)
 */

export function haversineM(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number
): number {
  const R = 6371000;
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLon / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

export interface SegmentDef {
  id: string;
  route: string;
  directionId: number;
  segmentIndex: number;
  startLat: number;
  startLon: number;
  endLat: number;
  endLon: number;
  midLat: number;
  midLon: number;
  lengthM: number;
  geometry: {
    type: "LineString";
    coordinates: [number, number][];
  };
}

export function splitIntoSegments(
  route: string,
  directionId: number,
  coordinates: [number, number][],
  targetLengthM: number = 200
): SegmentDef[] {
  if (coordinates.length < 2) return [];

  const segments: SegmentDef[] = [];
  let segmentIndex = 0;
  let segCoords: [number, number][] = [coordinates[0]];
  let segLength = 0;

  for (let i = 1; i < coordinates.length; i++) {
    const prev = coordinates[i - 1];
    const curr = coordinates[i];
    const dist = haversineM(prev[1], prev[0], curr[1], curr[0]);

    segCoords.push(curr);
    segLength += dist;

    if (segLength >= targetLengthM || i === coordinates.length - 1) {
      const start = segCoords[0];
      const end = segCoords[segCoords.length - 1];
      const midIdx = Math.floor(segCoords.length / 2);
      const mid = segCoords[midIdx];

      segments.push({
        id: `${route}:${directionId}:${segmentIndex}`,
        route,
        directionId,
        segmentIndex,
        startLat: start[1],
        startLon: start[0],
        endLat: end[1],
        endLon: end[0],
        midLat: mid[1],
        midLon: mid[0],
        lengthM: segLength,
        geometry: {
          type: "LineString",
          coordinates: segCoords.map((c) => [c[0], c[1]]),
        },
      });

      segmentIndex++;
      segCoords = [curr];
      segLength = 0;
    }
  }

  return segments;
}

export function snapToSegment(
  lat: number,
  lon: number,
  route: string,
  directionId: number | null,
  segments: SegmentDef[],
  maxDistM: number = 150
): string | null {
  let bestId: string | null = null;
  let bestDist = Infinity;

  for (const seg of segments) {
    if (seg.route !== route) continue;
    if (directionId !== null && seg.directionId !== directionId) continue;

    const dist = haversineM(lat, lon, seg.midLat, seg.midLon);
    if (dist < bestDist && dist <= maxDistM) {
      bestDist = dist;
      bestId = seg.id;
    }
  }

  return bestId;
}

File: worker/segments.go
package main

import (
	"fmt"
	"math"
)

// haversineM computes distance in meters between two lat/lon points
func haversineM(lat1, lon1, lat2, lon2 float64) float64 {
	const R = 6371000.0
	dLat := (lat2 - lat1) * math.Pi / 180
	dLon := (lon2 - lon1) * math.Pi / 180
	a := math.Sin(dLat/2)*math.Sin(dLat/2) +
		math.Cos(lat1*math.Pi/180)*math.Cos(lat2*math.Pi/180)*
			math.Sin(dLon/2)*math.Sin(dLon/2)
	return R * 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
}

// SegmentDef defines a route segment
type SegmentDef struct {
	ID           string
	Route        string
	DirectionID  int
	SegmentIndex int
	StartLat     float64
	StartLon     float64
	EndLat       float64
	EndLon       float64
	MidLat       float64
	MidLon       float64
	LengthM      float64
	Geometry     SegmentGeometry
}

type SegmentGeometry struct {
	Type        string       `json:"type"`
	Coordinates [][2]float64 `json:"coordinates"`
}

func splitIntoSegments(route string, directionID int, coordinates [][2]float64, targetLengthM float64) []SegmentDef {
	if len(coordinates) < 2 {
		return nil
	}

	var segments []SegmentDef
	segmentIndex := 0
	segCoords := [][2]float64{coordinates[0]}
	var segLength float64

	for i := 1; i < len(coordinates); i++ {
		prev := coordinates[i-1]
		curr := coordinates[i]
		dist := haversineM(prev[1], prev[0], curr[1], curr[0])

		segCoords = append(segCoords, curr)
		segLength += dist

		if segLength >= targetLengthM || i == len(coordinates)-1 {
			start := segCoords[0]
			end := segCoords[len(segCoords)-1]
			midIdx := len(segCoords) / 2
			mid := segCoords[midIdx]

			geomCoords := make([][2]float64, len(segCoords))
			copy(geomCoords, segCoords)

			segments = append(segments, SegmentDef{
				ID:           fmt.Sprintf("%s:%d:%d", route, directionID, segmentIndex),
				Route:        route,
				DirectionID:  directionID,
				SegmentIndex: segmentIndex,
				StartLat:     start[1],
				StartLon:     start[0],
				EndLat:       end[1],
				EndLon:       end[0],
				MidLat:       mid[1],
				MidLon:       mid[0],
				LengthM:      segLength,
				Geometry: SegmentGeometry{
					Type:        "LineString",
					Coordinates: geomCoords,
				},
			})

			segmentIndex++
			segCoords = [][2]float64{curr}
			segLength = 0
		}
	}

	return segments
}

func snapToSegment(lat, lon float64, route string, directionID *int16, segments []SegmentDef, maxDistM float64) string {
	bestID := ""
	bestDist := math.Inf(1)

	for i := range segments {
		seg := &segments[i]
		if seg.Route != route {
			continue
		}
		if directionID != nil && seg.DirectionID != int(*directionID) {
			continue
		}

		dist := haversineM(lat, lon, seg.MidLat, seg.MidLon)
		if dist < bestDist && dist <= maxDistM {
			bestDist = dist
			bestID = seg.ID
		}
	}

	return bestID
}
